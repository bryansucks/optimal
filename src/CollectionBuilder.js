/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import Builder from './Builder';

export default class CollectionBuilder<T, TDefault> extends Builder<?TDefault> {
  contents: ?Builder<T>;

  constructor(
    type: 'array' | 'object',
    contents?: Builder<T>,
    defaultValue?: ?TDefault = null,
  ) {
    super(type, defaultValue);

    if (contents) {
      if (contents instanceof Builder) {
        this.contents = contents;
        this.addCheck(this.checkContents, contents);

      } else if (__DEV__) {
        this.invariant(false, `A blueprint is required for ${type} contents.`);
      }
    }
  }

  checkContents(path: string, value: *, contents: Builder<T>) {
    if (__DEV__) {
      if (this.type === 'array') {
        value.forEach((item, i) => {
          contents.runChecks(
            `${path}[${i}]`,
            item,
            this.currentOptions,
            this.currentConfig,
          );
        });

      } else if (this.type === 'object') {
        Object.keys(value).forEach((key) => {
          contents.runChecks(
            `${path}.${key}`,
            value[key],
            this.currentOptions,
            this.currentConfig,
          );
        });
      }
    }
  }

  notEmpty(): this {
    return this.addCheck(this.checkNotEmpty);
  }

  checkNotEmpty(path: string, value: *) {
    if (__DEV__) {
      if (this.type === 'array') {
        this.invariant(
          (value.length > 0),
          'Array cannot be empty.',
          path,
        );

      } else if (this.type === 'object') {
        this.invariant(
          (Object.keys(value).length > 0),
          'Object cannot be empty.',
          path,
        );
      }
    }
  }

  /**
   * If contents are defined, return the type name using generics syntax.
   */
  typeAlias(): string {
    const { contents } = this;
    const alias = super.typeAlias();

    return contents ? `${alias}<${contents.typeAlias()}>` : alias;
  }
}

export function array<T>(
  contents?: Builder<T>,
  defaultValue?: ?T[] = [],
): CollectionBuilder<T, T[]> {
  return new CollectionBuilder('array', contents, defaultValue);
}

export function object<T>(
  contents?: Builder<T>,
  defaultValue?: ?{ [key: string]: T } = {},
): CollectionBuilder<T, { [key: string]: T }> {
  return new CollectionBuilder('object', contents, defaultValue);
}
