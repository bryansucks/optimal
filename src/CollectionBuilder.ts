import Builder from './Builder';

export default class CollectionBuilder<T> extends Builder<T> {
  sizeOf(length: number): this {
    if (__DEV__) {
      this.invariant(
        typeof length === 'number' && length > 0,
        'Size requires a non-zero positive number.',
      );
    }

    return this.addCheck(this.checkSizeOf, length);
  }

  checkSizeOf(path: string, value: T, length: number) {
    if (__DEV__) {
      // Array
      if (Array.isArray(value)) {
        this.invariant(value.length === length, `Array length must be ${length}.`, path);

        // String
      } else if (typeof value === 'string') {
        this.invariant(value.length === length, `String length must be ${length}.`, path);

        // Object
      } else if (typeof value === 'object' && value) {
        this.invariant(
          Object.keys(value).length === length,
          `Object must have ${length} properties.`,
          path,
        );

        // Unknown
      } else {
        this.invariant(false, 'Unknown type for size of checks.');
      }
    }
  }
}
