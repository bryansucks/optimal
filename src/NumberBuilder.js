/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import Builder from './Builder';

function isNumber(value: *): boolean {
  return (typeof value === 'number');
}

export default class NumberBuilder extends Builder<number> {
  constructor(defaultValue: number = 0) {
    super('number', defaultValue);
  }

  between(min: number, max: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(
        (isNumber(min) && isNumber(max)),
        'number.between() requires a minimum and maximum number.',
      );
    }

    return this.addCheck(this.checkBetween, min, max, inclusive);
  }

  checkBetween(path: string, value: *, min: number, max: number, inclusive: boolean = false) {
    if (__DEV__) {
      this.invariant(
        (
          isNumber(value) &&
          (inclusive ? (value >= min && value <= max) : (value > min && value < max))
        ),
        `Number must be between ${min} and ${max}${inclusive ? ' inclusive' : ''}.`,
        path,
      );
    }
  }

  oneOf(list: string[] = []): this {
    if (__DEV__) {
      this.invariant(
        (Array.isArray(list) && list.length > 0 && list.every(isNumber)),
        'number.oneOf() requires a non-empty array of numbers.',
      );
    }

    return this.addCheck(this.checkOneOf, list);
  }

  checkOneOf(path: string, value: *, list: string[] = []) {
    if (__DEV__) {
      this.invariant(
        (list.indexOf(value) >= 0),
        `Number must be one of: ${list.join(', ')}`,
        path,
      );
    }
  }
}

export function number(defaultValue: number = 0): NumberBuilder {
  return new NumberBuilder(defaultValue);
}
