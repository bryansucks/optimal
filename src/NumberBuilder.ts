/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Builder from './Builder';

function isNumber(value: any): boolean {
  return typeof value === 'number';
}

export default class NumberBuilder extends Builder<number | null> {
  constructor(defaultValue: number | null = 0) {
    super('number', defaultValue);
  }

  between(min: number, max: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(
        isNumber(min) && isNumber(max),
        'Between requires a minimum and maximum number.',
      );
    }

    return this.addCheck(this.checkBetween, min, max, inclusive);
  }

  checkBetween(path: string, value: any, min: number, max: number, inclusive: boolean = false) {
    if (__DEV__) {
      this.invariant(
        isNumber(value) && (inclusive ? value >= min && value <= max : value > min && value < max),
        `Number must be between ${min} and ${max}${inclusive ? ' inclusive' : ''}.`,
        path,
      );
    }
  }

  gt(min: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(isNumber(min), 'Greater-than requires a minimum number.');
    }

    return this.addCheck(this.checkGreaterThan, min, inclusive);
  }

  gte(min: number): this {
    return this.gt(min, true);
  }

  checkGreaterThan(path: string, value: any, min: number, inclusive: boolean = false) {
    if (__DEV__) {
      if (inclusive) {
        this.invariant(
          isNumber(value) && value >= min,
          `Number must be greater than or equal to ${min}.`,
          path,
        );
      } else {
        this.invariant(isNumber(value) && value > min, `Number must be greater than ${min}.`, path);
      }
    }
  }

  lt(max: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(isNumber(max), 'Less-than requires a maximum number.');
    }

    return this.addCheck(this.checkLessThan, max, inclusive);
  }

  lte(max: number): this {
    return this.lt(max, true);
  }

  checkLessThan(path: string, value: any, max: number, inclusive: boolean = false) {
    if (__DEV__) {
      if (inclusive) {
        this.invariant(
          isNumber(value) && value <= max,
          `Number must be less than or equal to ${max}.`,
          path,
        );
      } else {
        this.invariant(isNumber(value) && value < max, `Number must be less than ${max}.`, path);
      }
    }
  }

  oneOf(list: number[]): this {
    if (__DEV__) {
      this.invariant(
        Array.isArray(list) && list.length > 0 && list.every(item => isNumber(item)),
        'One of requires a non-empty array of numbers.',
      );
    }

    return this.addCheck(this.checkOneOf, list);
  }

  checkOneOf(path: string, value: any, list: number[]) {
    if (__DEV__) {
      this.invariant(list.indexOf(value) >= 0, `Number must be one of: ${list.join(', ')}`, path);
    }
  }
}

export function number(defaultValue: number | null = 0): NumberBuilder {
  return new NumberBuilder(defaultValue);
}
