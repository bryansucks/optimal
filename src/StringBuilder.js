/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import Builder from './Builder';

function isString(value: *): boolean {
  return (typeof value === 'string' && value !== '');
}

export default class StringBuilder extends Builder<string> {
  allowEmpty: boolean = false;

  constructor(defaultValue: string = '') {
    super('string', defaultValue);

    // Check emptiness
    this.addCheck(this.checkNotEmpty);
  }

  contains(token: string, index: number = 0): this {
    if (__DEV__) {
      this.invariant(
        isString(token),
        'Contains requires a non-empty string.',
      );
    }

    return this.addCheck(this.checkContains, token, index);
  }

  checkContains(path: string, value: *, token: string, index: number = 0) {
    if (__DEV__) {
      this.invariant(
        (value.indexOf(token, index) >= 0),
        `String does not include "${token}".`,
        path,
      );
    }
  }

  match(pattern: RegExp): this {
    if (__DEV__) {
      this.invariant(
        (pattern instanceof RegExp),
        'Match requires a regular expression to match against.',
      );
    }

    return this.addCheck(this.checkMatch, pattern);
  }

  checkMatch(path: string, value: *, pattern: RegExp) {
    if (__DEV__) {
      this.invariant(
        value.match(pattern),
        `String does not match pattern "${pattern.source}".`,
        path,
      );
    }
  }

  empty(): this {
    this.allowEmpty = true;

    return this;
  }

  checkNotEmpty(path: string, value: *) {
    if (__DEV__) {
      if (!this.allowEmpty) {
        this.invariant(
          isString(value),
          'String cannot be empty.',
          path,
        );
      }
    }
  }

  oneOf(list: string[] = []): this {
    if (__DEV__) {
      this.invariant(
        (Array.isArray(list) && list.length > 0 && list.every(isString)),
        'One of requires a non-empty array of strings.',
      );
    }

    return this.addCheck(this.checkOneOf, list);
  }

  checkOneOf(path: string, value: *, list: string[] = []) {
    if (__DEV__) {
      this.invariant(
        (list.indexOf(value) >= 0),
        `String must be one of: ${list.join(', ')}`,
        path,
      );
    }
  }
}

export function string(defaultValue: string = ''): StringBuilder {
  return new StringBuilder(defaultValue);
}
