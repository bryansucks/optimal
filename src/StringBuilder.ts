/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Builder from './Builder';

function isString(value: any): value is string {
  return typeof value === 'string' && value !== '';
}

export default class StringBuilder<T extends string = string> extends Builder<T> {
  allowEmpty: boolean = false;

  constructor(defaultValue?: T) {
    super('string', defaultValue || ('' as T));

    // Not empty by default
    if (__DEV__) {
      this.addCheck(this.checkNotEmpty);
    }
  }

  contains(token: string, index: number = 0): this {
    if (__DEV__) {
      this.invariant(isString(token), 'Contains requires a non-empty string.');
    }

    return this.addCheck(this.checkContains, token, index);
  }

  checkContains(path: string, value: T, token: string, index: number = 0) {
    if (__DEV__) {
      this.invariant(value.indexOf(token, index) >= 0, `String does not include "${token}".`, path);
    }
  }

  match(pattern: RegExp): this {
    if (__DEV__) {
      this.invariant(
        pattern instanceof RegExp,
        'Match requires a regular expression to match against.',
      );
    }

    return this.addCheck(this.checkMatch, pattern);
  }

  checkMatch(path: string, value: T, pattern: RegExp) {
    if (__DEV__) {
      this.invariant(
        !!value.match(pattern),
        `String does not match pattern "${pattern.source}".`,
        path,
      );
    }
  }

  empty(): this {
    if (__DEV__) {
      this.allowEmpty = true;
    }

    return this;
  }

  checkNotEmpty(path: string, value: T) {
    if (__DEV__) {
      if (!this.allowEmpty) {
        this.invariant(isString(value), 'String cannot be empty.', path);
      }
    }
  }

  oneOf<U extends string>(list: U[]) /* refine */ {
    if (__DEV__) {
      this.invariant(
        Array.isArray(list) && list.length > 0 && list.every(item => isString(item)),
        'One of requires a non-empty array of strings.',
      );
    }

    this.addCheck(this.checkOneOf, list);

    return (this as any) as StringBuilder<U>;
  }

  checkOneOf(path: string, value: T, list: T[]) {
    if (__DEV__) {
      this.invariant(list.indexOf(value) >= 0, `String must be one of: ${list.join(', ')}`, path);
    }
  }
}

export function string<T extends string = string>(defaultValue?: T) /* infer */ {
  return new StringBuilder<T>(defaultValue);
}
