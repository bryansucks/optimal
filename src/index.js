/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { bool, custom, func } from './Builder';
import { array, object } from './CollectionBuilder';
import { instance, date, regex } from './InstanceBuilder';
import { number } from './NumberBuilder';
import { shape } from './ShapeBuilder';
import { string } from './StringBuilder';
import { union } from './UnionBuilder';

export {
  array,
  bool,
  custom,
  date,
  func,
  instance,
  number,
  object,
  regex,
  shape,
  string,
  union,
};
export { default } from './Options';
