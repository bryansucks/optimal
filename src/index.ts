/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import optimal from './optimal';
import Builder, { bool, custom, func } from './Builder';
import ArrayBuilder, { array } from './ArrayBuilder';
import InstanceBuilder, { instance, date, regex } from './InstanceBuilder';
import NumberBuilder, { number } from './NumberBuilder';
import ObjectBuilder, { object } from './ObjectBuilder';
import ShapeBuilder, { shape } from './ShapeBuilder';
import StringBuilder, { string } from './StringBuilder';
import UnionBuilder, { union } from './UnionBuilder';

export { array, bool, custom, date, func, instance, number, object, regex, shape, string, union };

export {
  Builder,
  ArrayBuilder,
  InstanceBuilder,
  NumberBuilder,
  ObjectBuilder,
  ShapeBuilder,
  StringBuilder,
  UnionBuilder,
};

export * from './types';

export default optimal;
