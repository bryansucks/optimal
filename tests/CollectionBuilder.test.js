import { array, object } from '../src/CollectionBuilder';
import { number } from '../src/NumberBuilder';
import { string } from '../src/StringBuilder';

describe('array()', () => {
  let builder;

  beforeEach(() => {
    builder = array(string(), []);
  });

  describe('constructor()', () => {
    it('errors if a non-builder is passed', () => {
      expect(() => {
        builder = array(123);
      }).toThrowError('A blueprint is required for array contents.');
    });

    it('doesnt error if a builder is not passed', () => {
      expect(() => {
        builder = array();
      }).not.toThrowError('A blueprint is required for array contents.');
    });

    it('doesnt error if a builder is passed', () => {
      expect(() => {
        builder = array(string());
      }).not.toThrowError('A blueprint is required for array contents.');
    });

    it('sets type and default value', () => {
      builder = array(string(), ['foo']);

      expect(builder.type).toBe('array');
      expect(builder.defaultValue).toEqual(['foo']);
    });
  });

  describe('runChecks()', () => {
    it('returns an empty array for no data', () => {
      expect(builder.runChecks('key')).toEqual([]);
    });

    it('errors if a non-array is passed', () => {
      expect(() => {
        builder.runChecks('key', 'foo');
      }).toThrowError('Invalid option "key". Must be an array.');
    });

    it('errors if a non-array is passed, when not using a builder', () => {
      expect(() => {
        array().runChecks('key', 'foo');
      }).toThrowError('Invalid option "key". Must be an array.');
    });

    it('checks each item in the array', () => {
      expect(() => {
        builder.runChecks('key', ['foo', 'bar', 'baz', 123]);
      }).toThrowError('Invalid option "key[3]". Must be a string.');
    });

    it('errors if an array item is invalid; persists path with index', () => {
      expect(() => {
        builder.runChecks('key', [123]);
      }).toThrowError('Invalid option "key[0]". Must be a string.');
    });

    it('supports arrays of arrays', () => {
      builder = array(array(string()));

      const data = [
        ['foo', 'bar'],
        ['baz', 'qux'],
      ];

      expect(builder.runChecks('key', data)).toEqual(data);
    });

    it('errors correctly for arrays in arrays', () => {
      builder = array(array(string()));

      expect(() => {
        builder.runChecks('key', [
          ['foo', 'bar'],
          ['baz', 123],
        ]);
      }).toThrowError('Invalid option "key[1][1]". Must be a string.');
    });
  });

  describe('notEmpty()', () => {
    it('adds a checker', () => {
      builder.notEmpty();

      expect(builder.checks[2]).toEqual({
        func: builder.checkNotEmpty,
        args: [],
      });
    });
  });

  describe('checkNotEmpty()', () => {
    it('errors if value is empty', () => {
      expect(() => {
        builder.checkNotEmpty('key', []);
      }).toThrowError('Invalid option "key". Array cannot be empty.');
    });

    it('doesnt error if not empty', () => {
      expect(() => {
        builder.checkNotEmpty('key', [123]);
      }).not.toThrowError('Invalid option "key". Array cannot be empty.');
    });
  });

  describe('typeAlias()', () => {
    it('returns the type name when no contents', () => {
      expect(array().typeAlias()).toBe('Array');
    });

    it('returns the type name with contents type', () => {
      expect(array(string()).typeAlias()).toBe('Array<String>');
    });
  });
});

describe('object()', () => {
  let builder;

  beforeEach(() => {
    builder = object(string(), {});
  });

  describe('constructor()', () => {
    it('errors if a non-builder is passed', () => {
      expect(() => {
        builder = object(123);
      }).toThrowError('A blueprint is required for object contents.');
    });

    it('doesnt error if a builder is not passed', () => {
      expect(() => {
        object();
      }).not.toThrowError('A blueprint is required for object contents.');
    });

    it('doesnt error if a builder is passed', () => {
      expect(() => {
        builder = object(string());
      }).not.toThrowError('A blueprint is required for object contents.');
    });

    it('sets type and default value', () => {
      builder = object(string(), { foo: 'bar' });

      expect(builder.type).toBe('object');
      expect(builder.defaultValue).toEqual({ foo: 'bar' });
    });
  });

  describe('runChecks()', () => {
    it('returns an empty object for no data', () => {
      expect(builder.runChecks('key')).toEqual({});
    });

    it('errors if a non-object is passed', () => {
      expect(() => {
        builder.runChecks('key', 'foo');
      }).toThrowError('Invalid option "key". Must be a plain object.');
    });

    it('errors if a non-object is passed, when not using a builder', () => {
      expect(() => {
        object().runChecks('key', 'foo');
      }).toThrowError('Invalid option "key". Must be a plain object.');
    });

    it('checks each item in the object', () => {
      expect(() => {
        builder.runChecks('key', {
          a: 'foo',
          b: 'bar',
          c: 123,
        });
      }).toThrowError('Invalid option "key.c". Must be a string.');
    });

    it('errors if an object item is invalid; persists path with index', () => {
      expect(() => {
        builder.runChecks('key', {
          foo: 123,
        });
      }).toThrowError('Invalid option "key.foo". Must be a string.');
    });

    it('supports objects of objects', () => {
      builder = object(object(string()));

      const data = {
        a: {
          foo: '123',
          bar: '456',
        },
        b: {
          baz: '789',
        },
      };

      expect(builder.runChecks('key', data)).toEqual(data);
    });

    it('errors correctly for objects in objects', () => {
      builder = object(object(string()));

      expect(() => {
        builder.runChecks('key', {
          a: {
            foo: '123',
            bar: 456,
          },
          b: {
            baz: '789',
          },
        });
      }).toThrowError('Invalid option "key.a.bar". Must be a string.');
    });
  });

  describe('notEmpty()', () => {
    it('adds a checker', () => {
      builder.notEmpty();

      expect(builder.checks[2]).toEqual({
        func: builder.checkNotEmpty,
        args: [],
      });
    });
  });

  describe('checkNotEmpty()', () => {
    it('errors if value is empty', () => {
      expect(() => {
        builder.checkNotEmpty('key', {});
      }).toThrowError('Invalid option "key". Object cannot be empty.');
    });

    it('doesnt error if not empty', () => {
      expect(() => {
        builder.checkNotEmpty('key', { foo: 123 });
      }).not.toThrowError('Invalid option "key". Object cannot be empty.');
    });
  });

  describe('typeAlias()', () => {
    it('returns the type name when no contents', () => {
      expect(object().typeAlias()).toBe('Object');
    });

    it('returns the type name with contents type', () => {
      expect(object(number()).typeAlias()).toBe('Object<Number>');
    });
  });
});
