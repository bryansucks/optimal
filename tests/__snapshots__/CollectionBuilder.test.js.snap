// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array() checkNotEmpty() errors if value is empty 1`] = `"Invalid field \\"key\\". Array cannot be empty."`;

exports[`array() constructor() errors if a non-builder is passed 1`] = `"A blueprint is required for array contents."`;

exports[`array() runChecks() checks each item in the array 1`] = `"Invalid field \\"key[3]\\". Must be a string."`;

exports[`array() runChecks() errors correctly for arrays in arrays 1`] = `"Invalid field \\"key[1][1]\\". Must be a string."`;

exports[`array() runChecks() errors if a non-array is passed 1`] = `"Invalid field \\"key\\". Must be an array."`;

exports[`array() runChecks() errors if a non-array is passed, when not using a builder 1`] = `"Invalid field \\"key\\". Must be an array."`;

exports[`array() runChecks() errors if an array item is invalid; persists path with index 1`] = `"Invalid field \\"key[0]\\". Must be a string."`;

exports[`object() checkNotEmpty() errors if value is empty 1`] = `"Invalid field \\"key\\". Object cannot be empty."`;

exports[`object() constructor() errors if a non-builder is passed 1`] = `"A blueprint is required for object contents."`;

exports[`object() runChecks() checks each item in the object 1`] = `"Invalid field \\"key.c\\". Must be a string."`;

exports[`object() runChecks() errors correctly for objects in objects 1`] = `"Invalid field \\"key.a.bar\\". Must be a string."`;

exports[`object() runChecks() errors if a non-object is passed 1`] = `"Invalid field \\"key\\". Must be a plain object."`;

exports[`object() runChecks() errors if a non-object is passed, when not using a builder 1`] = `"Invalid field \\"key\\". Must be a plain object."`;

exports[`object() runChecks() errors if an object item is invalid; persists path with index 1`] = `"Invalid field \\"key.foo\\". Must be a string."`;
